### DISCUSSION TITLE
Do people still need to build the FIPS version of binaries/images with COMPLIANCE_POLICY: 1-140-2?

### DISCUSSION THREAD

**Discussion by qudongfang** [istio/istio#50599] [General]
Regarding

> Added an environment variable COMPLIANCE_POLICY to Istio components for enforcing TLS restriction for compliance with FIPS. When set to fips-140-2 on the Istiod container, the Istio Proxy container, and all other Istio components, the TLS version is restricted to v1.2. The cipher suites are limited to a subset of ECDHE-ECDSA-AES128-GCM-SHA256, ECDHE-RSA-AES128-GCM-SHA256, ECDHE-ECDSA-AES256-GCM-SHA384, ECDHE-RSA-AES256-GCM-SHA384, and ECDH curves to P-256.
> 
> These restrictions apply on the following data paths:
> 
> mTLS communication between Envoy proxies;
> regular TLS on the downstream and the upstream of Envoy proxies (e.g. gateway);
> Google gRPC side requests from Envoy proxies (e.g. Stackdriver extensions);
> Istiod xDS server;
> Istiod injection and validation webhook servers.
> The restrictions are not applied on the following data paths:
> 
> Istiod to Kubernetes API server;
> JWK fetch from Istiod;
> Wasm image and URL fetch from Istio Proxy containers;
> ztunnel.
> Note that Istio injector will propagate the value of COMPLIANCE_POLICY to the injected proxy container, when set. ([Issue #49081](https://github.com/istio/istio/issues/49081))

Can you help to clarify if people still need to build the FIPS version of binaries/images with COMPLIANCE_POLICY: 1-140-2?

cc @kyessenov

**Comment by pmerrison** [comment]
Yes, people will still need to.  The source of truth for all this is [CMVP 4407](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/4407).  The short version is that for Istio to be FIPS compliant, two things are needed: the first is only to use TLS 1.2, and the second is to make sure that the build pipeline is configured according to the steps outlined in the security policy so that the BoringCrypto module is used correctly.

**Comment by rubel-ahammad** [comment]
@qudongfang, Hi, I am new to Istio service mesh and I have already set up Istio in Ambient Mode successfully. Now that I want make Istio FIPS compliant, could you please share with me the list of binaries/images that we need to build for FIPS with COMPLIANCE_POLICY?

**Reply by qudongfang** [reply]
We don't use ambient mode.

We built all images that are used with FIPS.

Sorry that I cannot share details but 
https://gist.github.com/jeesmon/00f479a896548aadc0b32ae184890b4e is a good starting point.

**Reply by pmerrison** [reply]
You should be aware that it's not currently possible to use Ambient mode in a FIPS compliant manner. The work required to allow TLS 1.3 connections in Istio is being discussed/tracked [here](https://github.com/istio/istio/issues/52926)

**Comment by jared-schmidt-niceincontact** [comment]
An updated policy on cryptographic modules was released a month ago (specifically Jan. 16, 2025) [FedRAMP Policy for Cryptographic Module Selection and Use](https://www.fedramp.gov/updates/docs/cryptographic-module/) that mentions the fact that using a more current version of cryptographic modules may be more secure than using outdated FIPS-validated versions. Therefore, with proper documentation, using newer cryptographic modules may be justifiable.  Any thoughts on how that would affect this?

**Reply by krinkinmu** [reply]
I had a similar question. Also in the spirit of relying on more recent things, can we also reconsider limiting to TLS1.2 in FIPS mode and allow TLS1.3 as well?

My understanding is that FIPS itself does not require limiting TLS to 1.2 and it's more of our decision on how to achieve FIPS compliance rather than a FIPS requirement. My understanding is that FIPS requires to limit cryptography to the NIST approved and FIPS validates algorithm implementations, but as long as you limit yourself to those algorithms, you're free to use TLS1.3 (and SP 800-53 even mandates support for TLS1.3, which suggests that TLS1.3 on top of FIPS-validated crypto is a thing).

So maybe besides moving crypto modules to update-stream we can also allow both TLS1.2 and TLS1.3 and instead rely on something like https://github.com/google/boringssl/blob/ec09a2dad2781e35bdb2a3b873ff395b4376e444/ssl/ssl_lib.cc#L3421-L3446 instead. I'm also happy to work on contributing the required changes to make that happen, if we can agree that it's desirable.

**Comment by howardjohn** [comment]
the reason 1.3 is turned off is the only fips validated boringssl libraries
are ones before the compliance policy was added.

That being said the new guidance may change things.

**Reply by krinkinmu** [reply]
Checking BoringSSL repo, the main branch (which I think what they intend to advertise as an update-stream, based on https://boringssl.googlesource.com/boringssl/+/refs/heads/main/crypto/fipsmodule/FIPS.md#update-stream) does include compliance policies now.

**Reply by howardjohn** [reply]
Yeah I just mean the current _validated_ libraries. The new guidance around 'update streams' is quite new so hasn't really been broadly digested yet from what I can tell (https://github.com/envoyproxy/envoy/issues/38353 etc)
